### PR TITLE
Allow Custom Dataset in RAG Retriever

### PR DESCRIPTION
**Pull request by lhoestq**
As asked in #7725 , #7462 and #7631 , I added a way to let users build and load their own knowledge dataset for RAG.
I also added an example script that shows how to do that from csv files.

Before merging I'd like to make sure it creates the exact same embeddings that the one that were computed by the RAG team. I might need to do adjustments to the tokenization and maybe change the DPR encoder from the one trained on Natural Questions to the one trained ont the multiset/hybrid dataset.

Any feedbacks on the example and the HFIndex changes are welcome !

More details about the changes:
Previously the HFIndex only allowed to load existing datasets ("canonical" datasets) from the datasets library. So I split it into two classes `CanonicalHFIndex` to load canonical datasets and `CustomHFIndex` for custom user-defined ones.
Moreover the `config.index_name` used to accept any canonical dataset name, or "legacy" for the index the RAG team first provided. Now `config.index_name` can also accept "custom" for custom user-defined indexed datasets.

### CODE REVIEW

File: examples/rag/use_own_knowledge_dataset.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+    # The dataset needed for RAG must have three columns:
+    # - title (string): title of the document
+    # - text (string): text of a passage of the document
+    # - embeddings (array of dimension d): DPR representation of the passage
+
+    # Let's say you have documents in tab-separated csv files with columns "title" and "text"
+    csv_path = str(Path(__file__).parent / "test_data" / "my_knowledge_dataset.csv")
+
+    # You can load a Dataset object this way
```
Comment (thomwolf) [inline examples/rag/use_own_knowledge_dataset.py:69]: Add a link to the CSV loading page in the datasets doc

File: examples/rag/use_own_knowledge_dataset.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    # And save the index
+    index_path = os.path.join(tmp_dir, "my_knowledge_dataset_hnsw_index.faiss")
+    dataset.get_index("embeddings").save(index_path)
+
+    ######################################
+    print("Step 3 - Load RAG")
+    ######################################
+
+    retriever = RagRetriever.from_pretrained(
+        rag_model_name, index_name="custom", dataset=dataset_path, index_path=index_path
```
Comment (thomwolf) [inline examples/rag/use_own_knowledge_dataset.py:106]: Could we directly load a dataset with an index and prevent the user from having to save and load the index separately?

**Review by thomwolf — COMMENTED** [review]
Looks really cool! added a few comments

File: examples/rag/use_own_knowledge_dataset.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+def main(
+    tmp_dir: str,
+    num_proc=2,
+    batch_size=16,
+    rag_model_name="facebook/rag-sequence-nq",
+    dpr_ctx_encoder_model_name="facebook/dpr-ctx_encoder-single-nq-base",
+):
+
+    ######################################
+    print("Step 1 - Create the dataset")
```
Comment (patrickvonplaten) [inline examples/rag/use_own_knowledge_dataset.py:58]: (nit) would prefer `logger.log(...)` here.

File: examples/rag/use_own_knowledge_dataset.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    ######################################
+
+    # The dataset needed for RAG must have three columns:
+    # - title (string): title of the document
+    # - text (string): text of a passage of the document
+    # - embeddings (array of dimension d): DPR representation of the passage
+
+    # Let's say you have documents in tab-separated csv files with columns "title" and "text"
+    csv_path = str(Path(__file__).parent / "test_data" / "my_knowledge_dataset.csv")
```
Comment (patrickvonplaten) [inline examples/rag/use_own_knowledge_dataset.py:67]: Can we put `csv_path` also in the main with the default "test_data/my_knowledge_dataset.csv"? => probs better than hardcoding it here

File: examples/rag/use_own_knowledge_dataset.py (hunk 2 above)
Comment (lhoestq) [reply examples/rag/use_own_knowledge_dataset.py:106]: Sure we can. In that case we can just save the dataset and the faiss index when calling `save_pretrained`.
I'll see what I can do

File: examples/rag/use_own_knowledge_dataset.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    # And finally save your dataset
+    dataset_path = os.path.join(tmp_dir, "my_knowledge_dataset")
+    dataset.save_to_disk(dataset_path)
+
+    ######################################
+    print("Step 2 - Index the dataset")
+    ######################################
+
+    # Let's use the Faiss implementation of HNSW for fast approximate nearest neighbor search
+    d = 768  # vectors dimension
```
Comment (patrickvonplaten) [inline examples/rag/use_own_knowledge_dataset.py:92]: Could we maybe create a `@dataclass` to store all the config params => we could leverage the `hfArgParser` here.

File: src/transformers/retrieval_rag.py [hunk 6]
```diff
@@ -190,53 +190,76 @@ def get_top_docs(self, question_hidden_states: np.ndarray, n_docs=5) -> Tuple[np
         return np.array(ids), np.array(vectors)
 
 
-class HFIndex:
+class HFIndexBase:
```
Comment (patrickvonplaten) [inline src/transformers/retrieval_rag.py:193]: That's slightly backwards breaking, but OK IMO since people probably didn't use the class directly

**Review by patrickvonplaten — APPROVED** [review]
This is great! Really looking forward to it :-) 

Think we could clean up `use_own_knowledge_dataset.py` a bit by using `@dataclass` + `HfArgumentParser`

File: examples/rag/use_own_knowledge_dataset.py (hunk 2 above)
Comment (lhoestq) [reply examples/rag/use_own_knowledge_dataset.py]: I added a way to pass the indexed dataset directly, and I made it compatible with save/load_pretrained

**Comment by lhoestq** [comment]
I took your comments into account, let me know if you have other things to improve.
Also I had to change the DPR encoder from the one trained on Natural Questions to the one trained ont the multiset/hybrid dataset to match the embeddings used by the Rag team.

**Comment by mchari** [comment]
can't wait to try this out. 
@lhoestq , can this functionality be adapted to address #6399 ?

**Comment by ioannist** [comment]
Gald this got implemented! Many thanks Quentin. I checked out a copy, added a custom 25MB CSV file, and gave it a run:
python examples/rag/use_own_knowledge_dataset.py

Got this. Any ideas?
```
Traceback (most recent call last):
  File "examples/rag/use_own_knowledge_dataset.py", line 195, in <module>
    main(tmp_dir, rag_example_args, processing_args, index_hnsw_args)
  File "examples/rag/use_own_knowledge_dataset.py", line 84, in main
    ctx_tokenizer = DPRContextEncoderTokenizerFast.from_pretrained(rag_example_args.dpr_ctx_encoder_model_name)
  File "/home/ioannis/anaconda3/envs/transformers/lib/python3.6/site-packages/transformers/tokenization_utils_base.py", line 1544, in from_pretrained
    list(cls.vocab_files_names.values()),
OSError: Model name 'facebook/dpr-ctx_encoder-multiset-base' was not found in tokenizers model name list (facebook/dpr-ctx_encoder-single-nq-base). We assumed 'facebook/dpr-ctx_encoder-multiset-base' was a path, a model identifier, or url to a directory containing vocabulary files named ['vocab.txt'] but couldn't find such vocabulary files at this path or url.
```

**Comment by Deepika-Jain** [comment]
> Glad this got implemented! Many thanks @lhoestq . I checked out a copy, added a custom 25MB CSV file, and gave it a run:
> python examples/rag/use_own_knowledge_dataset.py
> 
> Got this.
> 
> ```
> Traceback (most recent call last):
>   File "examples/rag/use_own_knowledge_dataset.py", line 195, in <module>
>     main(tmp_dir, rag_example_args, processing_args, index_hnsw_args)
>   File "examples/rag/use_own_knowledge_dataset.py", line 84, in main
>     ctx_tokenizer = DPRContextEncoderTokenizerFast.from_pretrained(rag_example_args.dpr_ctx_encoder_model_name)
>   File "/home/ioannis/anaconda3/envs/transformers/lib/python3.6/site-packages/transformers/tokenization_utils_base.py", line 1544, in from_pretrained
>     list(cls.vocab_files_names.values()),
> OSError: Model name 'facebook/dpr-ctx_encoder-multiset-base' was not found in tokenizers model name list (facebook/dpr-ctx_encoder-single-nq-base). We assumed 'facebook/dpr-ctx_encoder-multiset-base' was a path, a model identifier, or url to a directory containing vocabulary files named ['vocab.txt'] but couldn't find such vocabulary files at this path or url.
> ```
> 
> I then switched to **dpr-ctx_encoder-single-nq-base**. In this case, indexing was successful. However, after Step 3 - Load RAG, the script started downloading the 74G wiki dataset (which should not be necessary) and then errored out on line 195.
> 
> ```
> $ python examples/rag/use_own_knowledge_dataset.py
> INFO:__main__:Step 1 - Create the dataset
> Using custom data configuration default
> Reusing dataset csv (/home/ioannis/.cache/huggingface/datasets/csv/default-49c04e2dbd1cfa6f/0.0.0/49187751790fa4d820300fd4d0707896e5b941f1a9c644652645b866716a4ac4)
> 100%|█████████████████████████████████████████| 155/155 [00:01<00:00, 82.27ba/s]
> 100%|█████████████████████████████████████| 18531/18531 [28:23<00:00, 10.88ba/s]
> INFO:__main__:Step 2 - Index the dataset
> 100%|█████████████████████████████████████████| 297/297 [30:22<00:00,  6.14s/it]
> INFO:__main__:Step 3 - Load RAG
> Using custom data configuration psgs_w100.nq.no_index
> Downloading and preparing dataset wiki_dpr/psgs_w100.nq.no_index (download: Unknown size, generated: Unknown size, post-processed: Unknown size, total: Unknown size) to /home/ioannis/.cache/huggingface/datasets/wiki_dpr/psgs_w100.nq.no_index/0.0.0/14b973bf2a456087ff69c0fd34526684eed22e48e0dfce4338f9a22b965ce7c2...
> Downloading: 100%|█████████████████████████| 11.2k/11.2k [00:00<00:00, 2.65MB/s]
> Downloading: 100%|███████████████████████| 78.4G/78.4G [1:54:59<00:00, 11.4MB/s]
> Dataset wiki_dpr downloaded and prepared to /home/ioannis/.cache/huggingface/datasets/wiki_dpr/psgs_w100.nq.no_index/0.0.0/14b973bf2a456087ff69c0fd34526684eed22e48e0dfce4338f9a22b965ce7c2. Subsequent calls will reuse this data.
> Using custom data configuration psgs_w100.nq.custom
> Downloading and preparing dataset wiki_dpr/psgs_w100.nq.custom (download: Unknown size, generated: Unknown size, post-processed: Unknown size, total: Unknown size) to /home/ioannis/.cache/huggingface/datasets/wiki_dpr/psgs_w100.nq.custom/0.0.0/14b973bf2a456087ff69c0fd34526684eed22e48e0dfce4338f9a22b965ce7c2...
> huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
> To disable this warning, you can either:
> 	- Avoid using `tokenizers` before the fork if possible
> 	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
> huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
> ...
> ...
> Downloading: 100%|██████████████████████████| 1.33G/1.33G [29:21<00:00, 753kB/s]
> Downloading: 100%|██████████████████████████| 1.33G/1.33G [29:30<00:00, 749kB/s]
> ...
> ...
> Traceback (most recent call last):            
>   File "examples/rag/use_own_knowledge_dataset.py", line 195, in <module>
>     main(tmp_dir, rag_example_args, processing_args, index_hnsw_args)
>   File "examples/rag/use_own_knowledge_dataset.py", line 116, in main
>     rag_example_args.rag_model_name, index_name="custom", indexed_dataset=dataset
>   File "/home/ioannis/anaconda3/envs/transformers/lib/python3.6/site-packages/transformers/retrieval_rag.py", line 321, in from_pretrained
>     config, question_encoder_tokenizer=question_encoder_tokenizer, generator_tokenizer=generator_tokenizer
>   File "/home/ioannis/anaconda3/envs/transformers/lib/python3.6/site-packages/transformers/retrieval_rag.py", line 310, in __init__
>     self.init_retrieval()
>   File "/home/ioannis/anaconda3/envs/transformers/lib/python3.6/site-packages/transformers/retrieval_rag.py", line 338, in init_retrieval
>     self.index.init_index()
>   File "/home/ioannis/anaconda3/envs/transformers/lib/python3.6/site-packages/transformers/retrieval_rag.py", line 248, in init_index
>     dummy=self.use_dummy_dataset,
>   File "/home/ioannis/anaconda3/envs/transformers/lib/python3.6/site-packages/datasets/load.py", line 611, in load_dataset
>     ignore_verifications=ignore_verifications,
>   File "/home/ioannis/anaconda3/envs/transformers/lib/python3.6/site-packages/datasets/builder.py", line 476, in download_and_prepare
>     dl_manager=dl_manager, verify_infos=verify_infos, **download_and_prepare_kwargs
>   File "/home/ioannis/anaconda3/envs/transformers/lib/python3.6/site-packages/datasets/builder.py", line 553, in _download_and_prepare
>     self._prepare_split(split_generator, **prepare_split_kwargs)
>   File "/home/ioannis/anaconda3/envs/transformers/lib/python3.6/site-packages/datasets/builder.py", line 841, in _prepare_split
>     generator, unit=" examples", total=split_info.num_examples, leave=False, disable=not_verbose
>   File "/home/ioannis/anaconda3/envs/transformers/lib/python3.6/site-packages/tqdm/std.py", line 1133, in __iter__
>     for obj in iterable:
>   File "/home/ioannis/.cache/huggingface/modules/datasets_modules/datasets/wiki_dpr/14b973bf2a456087ff69c0fd34526684eed22e48e0dfce4338f9a22b965ce7c2/wiki_dpr.py", line 124, in _generate_examples
>     id, text, title = line.strip().split("\t")
> ValueError: not enough values to unpack (expected 3, got 2)
> ```

I am facing the same issue.

**Comment by lhoestq** [comment]
Thanks for reporting I'm looking into it

**Comment by lhoestq** [comment]
> Glad this got implemented! Many thanks @lhoestq . I checked out a copy, added a custom 25MB CSV file, and gave it a run:
> python examples/rag/use_own_knowledge_dataset.py
> 
> Got this.
> 
> ```
> OSError: Model name 'facebook/dpr-ctx_encoder-multiset-base' was not found in tokenizers model name list (facebook/dpr-ctx_encoder-single-nq-base). We assumed 'facebook/dpr-ctx_encoder-multiset-base' was a path, a model identifier, or url to a directory containing vocabulary files named ['vocab.txt'] but couldn't find such vocabulary files at this path or url.
> ```
> 
> I then switched to **dpr-ctx_encoder-single-nq-base**. In this case, indexing was successful. However, after Step 3 - Load RAG, the script started downloading the 74G wiki dataset (which should not be necessary) and then errored out on line 195.
> 
> ```
> $ python examples/rag/use_own_knowledge_dataset.py
> ...
> ValueError: not enough values to unpack (expected 3, got 2)
> ```

You're having this issue because you are running the script with a version of transformers that doesn't include the changes I had to make in this PR to support custom datasets. This PR not only adds an example script, but there are also changes to make it possible in the core code of the RAG retriever.

**Comment by shamanez** [comment]
> > Glad this got implemented! Many thanks @lhoestq . I checked out a copy, added a custom 25MB CSV file, and gave it a run:
> > python examples/rag/use_own_knowledge_dataset.py
> > Got this.
> > ```
> > OSError: Model name 'facebook/dpr-ctx_encoder-multiset-base' was not found in tokenizers model name list (facebook/dpr-ctx_encoder-single-nq-base). We assumed 'facebook/dpr-ctx_encoder-multiset-base' was a path, a model identifier, or url to a directory containing vocabulary files named ['vocab.txt'] but couldn't find such vocabulary files at this path or url.
> > ```
> > 
> > 
> > I then switched to **dpr-ctx_encoder-single-nq-base**. In this case, indexing was successful. However, after Step 3 - Load RAG, the script started downloading the 74G wiki dataset (which should not be necessary) and then errored out on line 195.
> > ```
> > $ python examples/rag/use_own_knowledge_dataset.py
> > ...
> > ValueError: not enough values to unpack (expected 3, got 2)
> > ```
> 
> You're having this issue because you are running the script with a version of transformers that doesn't include the changes I had to make in this PR to support custom datasets. This PR not only adds an example script, but there are also changes to make it possible in the core code of the RAG retriever.
> 
> Everything works fine if you have all the changes of this PR

Having a separate script to fine-tune with custom datasets would be super useful!

**Comment by lhoestq** [comment]
> Having a separate script to fine-tune with custom datasets would be super useful!

I am adding flags to the fine-tuning scripts to make it work with a custom retriever ;)

**Comment by ioannist** [comment]
> > Glad this got implemented! Many thanks @lhoestq . I checked out a copy, added a custom 25MB CSV file, and gave it a run:
> > python examples/rag/use_own_knowledge_dataset.py
> > Got this.
> > ```
> > OSError: Model name 'facebook/dpr-ctx_encoder-multiset-base' was not found in tokenizers model name list (facebook/dpr-ctx_encoder-single-nq-base). We assumed 'facebook/dpr-ctx_encoder-multiset-base' was a path, a model identifier, or url to a directory containing vocabulary files named ['vocab.txt'] but couldn't find such vocabulary files at this path or url.
> > ```
> > 
> > 
> > I then switched to **dpr-ctx_encoder-single-nq-base**. In this case, indexing was successful. However, after Step 3 - Load RAG, the script started downloading the 74G wiki dataset (which should not be necessary) and then errored out on line 195.
> > ```
> > $ python examples/rag/use_own_knowledge_dataset.py
> > ...
> > ValueError: not enough values to unpack (expected 3, got 2)
> > ```
> 
> You're having this issue because you are running the script with a version of transformers that doesn't include the changes I had to make in this PR to support custom datasets. This PR not only adds an example script, but there are also changes to make it possible in the core code of the RAG retriever.
> 
> Everything works fine if you have all the changes of this PR

I just checkd out 14420dcbfa0f06012152cc66a7af84d7165a2a17 (committed one hour ago) and I am still getting the same error. Confused :)

`OSError: Model name 'facebook/dpr-ctx_encoder-multiset-base' was not found in tokenizers model name list (facebook/dpr-ctx_encoder-single-nq-base). We assumed 'facebook/dpr-ctx_encoder-multiset-base' was a path, a model identifier, or url to a directory containing vocabulary files named ['vocab.txt'] but couldn't find such vocabulary files at this path or url.`

**Comment by ioannist** [comment]
Ah, I think my mistake was that I was using the previous conda environment of transformers (with the new branch). Trying this again with a new env now :)

**Comment by lhoestq** [comment]
> I just checkd out [14420dc](https://github.com/huggingface/transformers/commit/14420dcbfa0f06012152cc66a7af84d7165a2a17) (committed one hour ago) and I am still getting the same error. Also, it was this branch that was giving the previous errors, not the main transformers branch. Perhaps some of your files have not been updated in the repository? Confused :)
> 
> `OSError: Model name 'facebook/dpr-ctx_encoder-multiset-base' was not found in tokenizers model name list (facebook/dpr-ctx_encoder-single-nq-base). We assumed 'facebook/dpr-ctx_encoder-multiset-base' was a path, a model identifier, or url to a directory containing vocabulary files named ['vocab.txt'] but couldn't find such vocabulary files at this path or url.`

Could you try to run
```python
from transformers import DPRContextEncoderTokenizerFast
tokenizer = DPRContextEncoderTokenizerFast.from_pretrained("facebook/dpr-ctx_encoder-single-nq-base")
```
and let me know if you're having the OSError on 14420dcbfa0f06012152cc66a7af84d7165a2a17 ?

Also just to help me fix this issue, could you also tell me the input of this code please:
```python
from transformers.tokenization_dpr_fast import CONTEXT_ENCODER_PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES
print(CONTEXT_ENCODER_PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES)
# it should print {'facebook/dpr-ctx_encoder-single-nq-base': 512, 'facebook/dpr-ctx_encoder-multiset-base': 512}
```

**Comment by ioannist** [comment]
> > I just checkd out [14420dc](https://github.com/huggingface/transformers/commit/14420dcbfa0f06012152cc66a7af84d7165a2a17) (committed one hour ago) and I am still getting the same error. Also, it was this branch that was giving the previous errors, not the main transformers branch. Perhaps some of your files have not been updated in the repository? Confused :)
> > `OSError: Model name 'facebook/dpr-ctx_encoder-multiset-base' was not found in tokenizers model name list (facebook/dpr-ctx_encoder-single-nq-base). We assumed 'facebook/dpr-ctx_encoder-multiset-base' was a path, a model identifier, or url to a directory containing vocabulary files named ['vocab.txt'] but couldn't find such vocabulary files at this path or url.`
> 
> Could you try to run
> 
> ```python
> from transformers import DPRContextEncoderTokenizerFast
> tokenizer = DPRContextEncoderTokenizerFast.from_pretrained("facebook/dpr-ctx_encoder-single-nq-base")
> ```
> 
> and let me know if you're having the OSError on [14420dc](https://github.com/huggingface/transformers/commit/14420dcbfa0f06012152cc66a7af84d7165a2a17) ?
> 
> Also just to help me fix this issue, could you also tell me the input of this code please:
> 
> ```python
> from transformers.tokenization_dpr_fast import CONTEXT_ENCODER_PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES
> print(CONTEXT_ENCODER_PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES)
> # it should print {'facebook/dpr-ctx_encoder-single-nq-base': 512, 'facebook/dpr-ctx_encoder-multiset-base': 512}
> ```

I was using the old transformers conda env. Doing things over again now and will report back asap!

**Comment by lhoestq** [comment]
Ok thanks !
I deleted my comment since it just noticed yours about the wrong env.
It should work for you now in a new env

**Comment by lhoestq** [comment]
I added some tests for the distributed retriever for fine-tuning.
Unless you still have an issue @ioannist this PR should be ready.
Cc @patrickvonplaten if you want to take a look at the new changes

**Review by patrickvonplaten — APPROVED** [review]
If the RAG slow tests all pass this PR is good to go for me

**Comment by ioannist** [comment]
> Ok thanks !
> I deleted my comment since I just noticed yours about the wrong env.
> It should work for you now in a new env

I created a new conda env from 14420dcb, installed pytorch and tf 2.2 (2.0 did not work). I ran again into the OS error. Here is the output on the tests you asked me to run +1 more.

(HEAD detached at 14420dcb)

```
from transformers import DPRContextEncoderTokenizerFast
tokenizer = DPRContextEncoderTokenizerFast.from_pretrained("facebook/dpr-ctx_encoder-single-nq-base")
```
No error

```
from transformers.tokenization_dpr_fast import CONTEXT_ENCODER_PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES
print(CONTEXT_ENCODER_PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES)
```
ModuleNotFoundError: No module named 'transformers.tokenization_dpr_fast'

```
from transformers.tokenization_dpr import CONTEXT_ENCODER_PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES
print(CONTEXT_ENCODER_PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES)
```
{'facebook/dpr-ctx_encoder-single-nq-base': 512}

**Comment by ioannist** [comment]
Give me an hour to test this more. I may be messing up somewhere and don't wanna waste your time..

**Comment by lhoestq** [comment]
If you can't import `transformers.tokenization_dpr_fast` that must be an environment issue. This file was added recently on the master branch. It looks like your env imports a version of transformers that is not up-to-date. That's why you're having an OSError.

Maybe you can run `pip show transformers` to check the location of your transformers installation ?

**Comment by ioannist** [comment]
> If you can't import `transformers.tokenization_dpr_fast` that must be an environment issue. This file was added recently on the master branch. It looks like your env imports a version of transformers that is not up-to-date. That's why you're having an OSError.
> 
> Maybe you can run `pip show transformers` to check the location of your transformers installation ?

Ok, I just did the env again for the 3rd time and it works! No idea what i messed up before. 

Step -3 Load Rag in progress :)

**Comment by lhoestq** [comment]
Slow tests pass. This one is ready to merge.

**Comment by patrickvonplaten** [comment]
@lhoestq - feel free to merge whenever!

**Comment by mchari** [comment]
When I do "from transformers import DPRContextEncoder", I get an error :
File "convert_slow_tokenizer.py", line 24, in <module>
    from tokenizers.models import BPE, Unigram, WordPiece
ImportError: cannot import name 'Unigram'

Unigram is missing in tokenizers 0.8.1.rc2. Needed to update to 0.9.2

**Comment by shamanez** [comment]
> I took your comments into account, let me know if you have other things to improve.
> Also I had to change the DPR encoder from the one trained on Natural Questions to the one trained ont the multiset/hybrid dataset to match the embeddings used by the Rag team.

@lhoestq 

Hi, can you elaborate on the change you made in the DPR bit more? My understanding is, you have pretrained the DPR with a hybrid dataset to improve the performance when encoding a custom knowledge base. 

If you have pretrained the DPR, can you please publish the code?

**Comment by lhoestq** [comment]
There are two versions of DPR in the paper. One trained on NQ and one trained on various datasets. The authors released the code and the weights in this [this repo](https://github.com/facebookresearch/DPR).

The change I did was just to use the weight of the second one, since it was the one used for RAG.

**Comment by shamanez** [comment]
Thanks. So as mentioned in the RAG paper, we can use the doc encoder to get
embeddings for any custom dataset.  Later we can only fine tune the BART
model and the question encoder.

**Comment by lhoestq** [comment]
Yes exactly :)

**Comment by mchari** [comment]
@lhoestq , I tried use_own_knowledge_dataset.py to try and retrieve passages from a custom dataset in response to queries. It works, though the relevance of the results isn't great. 
I wanted to try and use finetune.sh. Is it possible to finetune just the retriever ? Is there a sample format for the training data ? Looks like it will need the question and passages as input. Need positive and negative passages ? 
Thanks!

**Comment by shamanez** [comment]
@mchari  He previously said, it is possible to retrain DPR with Facebook code and then convert the checkpoint to the huge face compatible.

**Pull request merged**